code/ecdsap256sha2/Hacl.Impl.ECDSA.P256SHA256.Signature.Agile.fst

```diff
@@ -0,0 +1,319 @@
+
+let lemma_power_step6 kInv = 
+  let a = Spec.ECDSA.exponent_spec (fromDomain_ kInv) in 
+  lemmaFromDomain kInv;
+
+  power_distributivity (kInv * modp_inv2_prime (pow2 256) prime_p256_order) (prime_p256_order - 2) prime_p256_order;
+  power_distributivity_2 kInv (modp_inv2_prime (pow2 256) prime_p256_order % prime_p256_order) (prime_p256_order - 2);
+  lemma_mod_mul_distr_r (pow kInv (prime_p256_order - 2)) (pow (modp_inv2_prime (pow2 256) prime_p256_order) (prime_p256_order - 2)) prime_p256_order;
+
+  lemma_pow_mod_n_is_fpow prime_p256_order (pow2 256 % prime_p256_order) (prime_p256_order - 2);
+  
+  let inverse2_256 = 43790243014242295660885426880012836369732278457577312309071968676491870960761 in 
+  assert_norm(modp_inv2_prime (pow2 256) prime_p256_order = inverse2_256); 
+  lemma_pow_mod_n_is_fpow prime_p256_order inverse2_256 (prime_p256_order - 2);
+  assert_norm(exp #prime_p256_order inverse2_256 (prime_p256_order - 2) == pow2 256 % prime_p256_order);
+
+  lemma_mod_mul_distr_r (pow kInv (prime_p256_order - 2)) (pow2 256) prime_p256_order;
+  lemmaToDomain (pow kInv (prime_p256_order - 2))
+
+#push-options "--z3rlimit 300"
+
+inline_for_extraction
+val ecdsa_signature_step6: result: felem -> kFelem: felem -> z: felem -> r: felem -> da: felem -> Stack unit
+  (requires fun h -> 
+    live h result /\ live h kFelem /\ live h z /\ live h r /\ live h da /\
+    as_nat h kFelem < prime_p256_order /\ 
+    as_nat h z < prime_p256_order /\ 
+    as_nat h r < prime_p256_order /\ 
+    as_nat h da < prime_p256_order
+  )
+  (ensures fun h0 _ h1 -> 
+    modifies (loc result) h0 h1 /\
+    as_nat h1 result = (as_nat h0 z + as_nat h0 r * as_nat h0 da) * pow (as_nat h0 kFelem) (prime_p256_order - 2) % prime_p256_order
+  )
+
+let ecdsa_signature_step6 result kFelem z r da = 
+  let open FStar.Tactics in 
+  let open FStar.Tactics.Canon in 
+  push_frame();
+    let rda = create (size 4) (u64 0) in 
+    let zBuffer = create (size 4) (u64 0) in 
+    let kInv = create (size 4) (u64 0) in 
+  let h0 = ST.get() in 
+    montgomery_multiplication_ecdsa_module r da rda;
+    fromDomainImpl z zBuffer;
+    felem_add rda zBuffer zBuffer;  
+    copy kInv kFelem;
+    montgomery_ladder_exponent kInv;
+    montgomery_multiplication_ecdsa_module zBuffer kInv result;
+  pop_frame();
+      let br0 = as_nat h0 z + as_nat h0 r * as_nat h0 da in
+      let br1 = pow (as_nat h0 kFelem) (prime_p256_order - 2) in 
+
+      lemmaFromDomain (as_nat h0 r * as_nat h0 da); 
+      lemma_felem_add (as_nat h0 r * as_nat h0 da) (as_nat h0 z);
+      lemma_power_step6 (as_nat h0 kFelem); 
+
+      lemmaFromDomain (fromDomain_ br0);
+      lemmaToDomain br1;
+      assert_norm ((modp_inv2_prime (pow2 256) prime_p256_order * pow2 256) % prime_p256_order = 1);
+       
+      lemma_mod_mul_distr_l (fromDomain_ br0 * modp_inv2_prime (pow2 256) prime_p256_order) (br1 * pow2 256 % prime_p256_order) prime_p256_order;
```

Comment:
same thing about calc -- if this proof breaks and I need to fix it, I would have a much easier time with a `calc` statement